Do not raise error on unsupported plugins

filipesperandio opened the pull request:
Fix #192 

Instead of erroring and aborting the execution,
the engine now prints a warning and skip the plugin

Given an .eslintrc.yml:
```yaml
env:
    es6: true
    node: true
parserOptions:
    sourceType: module
plugins:
    - node
    - not_supported
extends:
    - not_valid
    - 'plugin:invalidplugin/recommended'
    - 'eslint:recommended'
    - 'plugin:node/recommended'
rules:
    invalidplugin/rule: 1
    node/exports-style: [error, module.exports]
    indent: [error, 4]
    linebreak-style: [error, unix]
    quotes: [error, double]
    semi: [error, always]
    no-console: off
```
When `CODECLIMATE_DEBUG=true codeclimate analyze` runs
Then the output is like:
```
...
Module not supported: eslint-plugin-not_supported
Module not supported: eslint-plugin-invalidplugin
Module not supported: eslint-config-not_valid
...
Analyzing: app1.js
...

== app1.js (1 issue) ==
1: Definition for rule 'invalidplugin/rule' was not found [eslint]
```

GordonDiggs commented on lib/eslint-patch.js:
```diff
@@ -1,33 +1,31 @@
+        Plugins.load(name);
+
+      } catch(e) {
+        console.error(`[DEBUG] Plugin ${name} not supported`);
```
Can we reword this as `Plugin not supported: ${name}` (dropping the `[DEBUG]` too)

filipesperandio replied:
Sure. The `[DEBUG]` idea was coming from other debug messages as they only get displayed with the debug flag on.

filipesperandio replied:
@GordonDiggs I've just noticed that in the issue #192 you mentioned we should tell the user we're *skipping* the plugin, should we add something on those lines?

GordonDiggs replied:
Eh I think this messaging is fine. Also, to be clear - this happens even if debug mode is disabled, right?

filipesperandio replied:
`CODECLIMATE_DEBUG=false` -> msg is still printed to stderr -> but CLI does not display it.

GordonDiggs replied:
@filipesperandio That's ok - it will always be visible on codeclimate.com

maxjacobson commented:
Did some local QA and determined that this does what it sets out to do: 🎉. One thing to consider is that when a user attempts to load a plugin, they're likely to also have their configuration set up to _use_ that plugin. For example, to enable a custom rule or to extend some configuration.

Here's what I observe:

* if I have an unrecognized rule enabled (like one defined in my plugin that was quietly not loaded):

```yaml
rules:
  - bogus/fafa: error
```

then _the engine runs, but it emits issues telling me that there's an unknown rule enabled_. This is OK but not great.

* If I try to extend an unknown config:

```yaml
extends:
  - bogus
```

then I get a crashing error:

```
Error: Cannot find module 'eslint-config-other-bogus
```

^ this one is a bigger problem IMO, since extending from configuration in packages is a common use case.

Those changes should probably be done as separate PRs, but I just wanted to note those. OK, going to take a closer look at the code now.

GordonDiggs commented:
@maxjacobson Filipe and I discussed that and decided that it's ok to to get the issues emitted about the missing rules. It may even be preferable as a way to alert users of an issue.

What's more, I believe that is the behavior of the `eslint-2` channel, which is a good basis

maxjacobson commented:
@GordonDiggs OK - that sounds good to me. What do you think about the `extends` thing?

GordonDiggs commented:
ah @maxjacobson I didn't realize you were saying that was the behavior as of this PR. @filipesperandio we should fix that before merging

maxjacobson commented on lib/eslint-patch.js:
```diff
@@ -1,33 +1,31 @@
-
-module.exports = function patcher(eslint) {
-
-  meld.around(eslint.CLIEngine, 'loadPlugins', function(joinPoint) {
```
this was removed because this hook doesn't ever actually fire? (I think I saw another comment you made about this)

filipesperandio replied:
@maxjacobson The function is only implemented on eslint 1.9 and lower, so my understanding is it wasn't doing anything.

filipesperandio replied:
It never fired on my tests...

filipesperandio commented:
@maxjacobson @GordonDiggs Thanks for the feedback. Will look into the `extends` thing.

filipesperandio commented on bin/eslint.js:
```diff
@@ -8,10 +8,11 @@ process.chdir(CODE_DIR);
+var eslint = require('../lib/eslint-patch')();
 
 var CLIEngine = eslint.CLIEngine;
-var docs = eslint.docs;
```
@GordonDiggs @maxjacobson I felt this was misplaced and was impacting the way we load/patch eslint, so I decoupled it from `eslint` instance. Haven't found any other use of `eslint.docs` so I think we are good, but wanted to hight light this.

GordonDiggs replied:
Cool. We use the docs for the issue contents - as long as those still work we're probably ok

maxjacobson replied:
Confirmed docs are still showing up correctly

maxjacobson commented on test/eslint-patch_test.js:
```diff
@@ -51,7 +55,20 @@ describe("eslint-patch", function() {
     });
+  });
+
+  describe("Avoid error processing pluging loading from extends configuration", function() {
```
typo: `pluging` => `plugins`

maxjacobson commented on lib/eslint-patch.js:
```diff
@@ -1,13 +1,22 @@
+  ModuleResolver.prototype.resolve = function(name, path) {
+    try {
+      return resolve.apply(this, [name, path]);
+    } catch(e) {
```
I think we should log something here

filipesperandio replied:
@maxjacobson Any suggestions about the message? I am feeling it would just duplicate what we do at [line 34](https://github.com/codeclimate/codeclimate-eslint/pull/204/files#diff-2daeb06199eb4f1fdf8cd0fb63c70b21R34).

maxjacobson replied:
I think using the same message sounds fine. Are you saying it's already being printed? I didn't see it when doing local QA, so I thought it wasn't included there

maxjacobson commented on lib/eslint-patch.js:
```diff
@@ -1,13 +1,22 @@
 const Plugins = require("eslint/lib/config/plugins");
+const ModuleResolver = require("eslint/lib/util/module-resolver");
+
+const ConfigFile = require("eslint/lib/config/config-file");
```
it looks like we aren't using this variable

1. do we need to require it?
1. I'm surprised we don't have an eslint check creating issue about this unused variable (we have [this](https://github.com/codeclimate/styleguide/pull/43) in our styleguide) -- is that misconfigured maybe?

maxjacobson commented on test/eslint-patch_test.js:
```diff
@@ -0,0 +1,74 @@
+    });
+  });
+
+  describe("Avoid error processing pluging loading from extends configuration", function() {
```
typo: `pluging`

filipesperandio commented on lib/eslint-patch.js:
```diff
@@ -1,33 +1,48 @@
+        return `${__dirname}/empty-plugin.js`;
+      }
+
+      throw e;
```
@maxjacobson Keeping track of skipped plugins, now. If an `extends` config is related to one of them, we also skip it here. `bogus` entry should still result in error, which is raised and propagated all the way.

maxjacobson commented:
OK taking another look now. Thanks for your patience

maxjacobson commented:
@filipesperandio I'm still getting errors when I attempt to extend a configuration:

```yaml
plugins:
  - bogus

extends:
  - hello-world
```

error:

```
Plugin not supported: bogus
/usr/src/app/node_modules/eslint/lib/config/config-file.js:401
            throw e;
            ^

Error: Cannot find module 'eslint-config-hello-world'
Referenced from: /code/.eslintrc.yml
    at ModuleResolver.resolve (/usr/src/app/node_modules/eslint/lib/util/module-resolver.js:74:19)
    at ModuleResolver.resolve (/usr/src/app/lib/eslint-patch.js:17:22)
    at resolve (/usr/src/app/node_modules/eslint/lib/config/config-file.js:478:33)
    at load (/usr/src/app/node_modules/eslint/lib/config/config-file.js:495:26)
    at configExtends.reduceRight.e (/usr/src/app/node_modules/eslint/lib/config/config-file.js:391:36)
    at Array.reduceRight (native)
    at applyExtends (/usr/src/app/node_modules/eslint/lib/config/config-file.js:362:28)
    at Object.load (/usr/src/app/node_modules/eslint/lib/config/config-file.js:529:22)
    at loadConfig (/usr/src/app/node_modules/eslint/lib/config.js:63:33)
    at getLocalConfig (/usr/src/app/node_modules/eslint/lib/config.js:130:29)
```

(note: it does catch-and-log the "bogus" plugin, but not the "hello-world" configuration)

filipesperandio commented:
@maxjacobson No worries, let's get this right.

The current behaviour is actually intentional. Note the *config* in `eslint-config-hello-world` meaning it is not actually coming from a plugin. I basically assumed we are skipping plugins, not configs (bogus configs still may crash). Maybe we are skipping both?
In the case of skipping both I should track what was already printed to `stderr` to avoid showing the same message over-and-over and the message should be something an npm module, not necessarily a plugin.

maxjacobson commented:
IMO we should skip both! Both are backed by modules which may or may not be
vendored, leading to people having errored builds when it's not available.
The goal here is to prevent errored builds when we haven't vendored a
package, right?
On Tue, Mar 14, 2017 at 13:48 Filipe Esperandio <notifications@github.com>
wrote:

> @maxjacobson <https://github.com/maxjacobson> No worries, let's get this
> right.
>
> The current behaviour is actually intentional. Note the *config* in
> eslint-config-hello-world meaning it is not actually coming from a
> plugin. I basically assumed we are skipping plugins, not configs (bogus
> configs still may crash). Maybe we are skipping both?
> In the case of skipping both I should track what was already printed to
> stderr to avoid showing the same message over-and-over and the message
> should be something an npm module, not necessarily a plugin.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/codeclimate/codeclimate-eslint/pull/204#issuecomment-286503993>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABWvm8sd5MSq81AUcbSKx5tFwEkJybqoks5rltLcgaJpZM4MYnT4>
> .
>
-- 
Max Jacobson, Software Engineer / max@codeclimate.com / We're Hiring!
<http://grnh.se/wbfx6j>

filipesperandio commented:
@maxjacobson Right!

The confusion I made was thinking about cases where users have an actually bogus setup (that fails even if their `package.json` match ours).
For example:

```yaml
plugins:
  - not_supported

extends:
  - actually_invalid_config
```

In that case we won't stop ESLint either because we will not differentiate a valid not supported module from an invalid config. Do we care? Maybe not, ESLint will run and find what it can.

maxjacobson commented:
Is that distinction (between a package that actually doesn't exist anywhere and a package that does exist, just not in our image) specifically related to "extends" or does it also apply to "plugins"? I feel like it also applies to "plugins", and we feel it's a best user experience to skip over any missing package rather than erroring

filipesperandio commented:
@maxjacobson It is for both, yes. Agreed. Making the adjustments.

filipesperandio commented:
@maxjacobson @GordonDiggs 
I've changed the wording a bit to account for `extends` and `plugins` (basically node modules), not only `plugins`:
`Plugin not supported: ${name}` => `Module not supported: ${name}`

Added the current output I get to the top [comment](https://github.com/codeclimate/codeclimate-eslint/pull/204#issue-213160863)

GordonDiggs commented:
I'm cool with that wording

maxjacobson commented:
Same. This LGTM to me now!

lucasmazza commented on lib/eslint-patch.js:
```diff
@@ -1,33 +1,50 @@
+  Plugins.loadAll = function(pluginNames) {
+    const loadedPlugins = Object.keys(Plugins.getAll());
+    if(loadedPlugins.length > 0) {
+      return;
```
This might be a wild guess, but it feels like this early return is the culprit of #214. If `loadAll` is called multiple times with different arguments then only the first invocation does something, while the others won't do anything (and maybe expected plugins won't be loaded/defined anymore), while the original implementation on ESLint does not checks for `getAll` and always attempt  to `load` each given plugin. 

Maybe we can monkey patch this again with only the `try`/`catch` around each `load` call and see how it goes. 🤔

lucasmazza replied:
Other alternative would be to only patch `Plugins.load(pluginName)`, which is where the `require` we want to wrap happens. https://github.com/eslint/eslint/blob/7d1af86851bb4af03ed05515ffc5e59e5c0cda52/lib/config/plugins.js#L127-L141

filipesperandio replied:
@lucasmazza On it! :)
